src/coreclr/jit/valuenum.cpp

```diff
@@ -3322,26 +3322,32 @@ bool ValueNumStore::VNEvalShouldFold(var_types typ, VNFunc func, ValueNum arg0VN
         }
     }
 
-    // Is this a checked cast that will always throw an exception?
-    if (func == VNF_CastOvf)
+    // Is this a checked cast that will always throw an exception or one with an implementation-defined result?
+    if (VNFuncIsNumericCast(func))
     {
-        var_types castToType;
-        bool      fromUnsigned;
-        GetCastOperFromVN(arg1VN, &castToType, &fromUnsigned);
         var_types castFromType = TypeOfVN(arg0VN);
 
-        switch (castFromType)
+        // By policy, we do not fold overflowing conversions from floating-point types,
+        // as the result the C++ compiler gives us does not always match our own codegen.
+        if ((func == VNF_CastOvf) || varTypeIsFloating(castFromType))
```

Comment:
Not sure if I understand this - The comment says that "we do not fold overflowing conversions from floating-point types" , then why have we included `|| varTypeIsFloating(castFromType)` for `VNF_Cast`?

Reply:
Hmm, I suppose this is poor phrasing on my part... The intent of the comment was to say that `we don't fold *any* casts from floating-point types that overflow`, but I see how that's conflating the notion of "does the cast actually overflow" with "is this a `VNF_CastOvf`".

Will reword to `conversions from floating-point types that result in overflow`.